src/Algebra/Graph/NonEmpty/AdjacencyIntMap.hs

```diff
@@ -0,0 +1,716 @@
+
+@'vertex' 1 < 'vertex' 2
+'vertex' 3 < 'edge' 1 2
+'vertex' 1 < 'edge' 1 1
+'edge' 1 1 < 'edge' 1 2
+'edge' 1 2 < 'edge' 1 1 + 'edge' 2 2
+'edge' 1 2 < 'edge' 1 3@
+
+Note that the resulting order refines the
+'isSubgraphOf' relation and is compatible
+with 'overlay' and
+'connect' operations:
+
+@'isSubgraphOf' x y ==> x <= y@
+
+@x     <= x + y
+x + y <= x * y@
+-}
+newtype AdjacencyIntMap = NAIM { am :: AIM.AdjacencyIntMap }
+    deriving (Eq, Generic, NFData, Ord)
+
+-- | __Note:__ this does not satisfy the usual ring laws; see 'AdjacencyIntMap' for
+-- more details.
+instance Num AdjacencyIntMap where
+    fromInteger = vertex . fromInteger
+    (+)         = overlay
+    (*)         = connect
+    signum      = error "NonEmpty.AdjacencyIntMap.signum cannot be implemented."
+    abs         = id
+    negate      = id
+
+instance Show AdjacencyIntMap where
+    showsPrec p nam
+        | null vs    = error "NonEmpty.AdjacencyIntMap.Show: Graph is empty"
+        | null es    = showParen (p > 10) $ vshow vs
+        | vs == used = showParen (p > 10) $ eshow es
+        | otherwise  = showParen (p > 10) $
+                           showString "overlay (" . vshow (vs \\ used) .
+                           showString ") (" . eshow es . showString ")"
+      where
+        vs             = toList (vertexList1 nam)
+        es             = edgeList nam
+        vshow [x]      = showString "vertex "    . showsPrec 11 x
+        vshow xs       = showString "vertices1 " . showsPrec 11 xs
+        eshow [(x, y)] = showString "edge "      . showsPrec 11 x .
+                         showString " "          . showsPrec 11 y
+        eshow xs       = showString "edges1 "    . showsPrec 11 xs
+        used           = IntSet.toAscList $ IntSet.fromList $ uncurry (++) $ unzip es
+
+-- Unsafe creation of a NonEmpty list.
+unsafeNonEmpty :: [Int] -> NonEmpty Int
+unsafeNonEmpty = fromMaybe (error msg) . nonEmpty
+  where
+    msg = "Algebra.Graph.AdjacencyIntMap.unsafeNonEmpty: Graph is empty"
+
+-- | Convert a possibly empty 'AIM.AdjacencyIntMap' into NonEmpty.'AdjacencyIntMap'.
+-- Returns 'Nothing' if the argument is 'AIM.empty'.
+-- Complexity: /O(1)/ time, memory and size.
+--
+-- @
+-- toNonEmpty 'AIM.empty'          == 'Nothing'
+-- toNonEmpty . 'fromNonEmpty' == 'Just'
+-- @
+toNonEmpty :: AIM.AdjacencyIntMap -> Maybe AdjacencyIntMap
+toNonEmpty x | AIM.isEmpty x = Nothing
+             | otherwise    = Just (NAIM x)
+
+-- | Convert a NonEmpty.'AdjacencyIntMap' into an 'AIM.AdjacencyIntMap'. The resulting
+-- graph is guaranteed to be non-empty.
+-- Complexity: /O(1)/ time, memory and size.
+--
+-- @
+-- 'isEmpty' . fromNonEmpty    == 'const' 'False'
+-- 'toNonEmpty' . fromNonEmpty == 'Just'
+-- @
+fromNonEmpty :: AdjacencyIntMap -> AIM.AdjacencyIntMap
+fromNonEmpty = am
+
+-- | Construct the graph comprising /a single isolated vertex/.
+-- Complexity: /O(1)/ time and memory.
+--
+-- @
+-- 'AdjacencyIntMap.hasVertex' x (vertex x) == True
```

Comment:
I'm a bit confused: do we really need qualifications `AdjacencyIntMap.` here?

Reply:
I don't think so. I was following very closely the `NonEmpty.AdjacencyMap` module

Reply:
Ha, indeed. If it's redundant, let's drop it there too.